src/MongoDB.Driver.Core/Core/WireProtocol/CommandUsingCommandMessageWireProtocol.cs

```diff
@@ -262,54 +262,81 @@ private Type0CommandMessageSection<BsonDocument> CreateType0Section(ConnectionDe
         {
             var extraElements = new List<BsonElement>();
 
-            addIfNotAlreadyAdded("$db", _databaseNamespace.DatabaseName);
+            AddIfNotAlreadyAdded("$db", _databaseNamespace.DatabaseName);
 
             if (connectionDescription.IsMasterResult.ServerType != ServerType.Standalone
                 && _readPreference != null
                 && _readPreference != ReadPreference.Primary)
             {
                 var readPreferenceDocument = QueryHelper.CreateReadPreferenceDocument(_readPreference);
-                addIfNotAlreadyAdded("$readPreference", readPreferenceDocument);
+                AddIfNotAlreadyAdded("$readPreference", readPreferenceDocument);
             }
 
             if (_session.Id != null)
             {
-                addIfNotAlreadyAdded("lsid", _session.Id);
+                if (IsSessionAcknowledged())
+                {
+                    AddIfNotAlreadyAdded("lsid", _session.Id);
+                }
+                else
+                {
+                    if (_session.IsImplicit)
+                    {
+                        // do not set sessionId If session is implicit and write is unacknowledged
+                    }
+                    else
+                    {
+                        throw new NotSupportedException("Explicit session must not be used with unacknowledged writes.");
```

Comment:
nit: I think this should probably be a `InvalidOperationException`.

Not supported and invalid operation are subtly different. Not supported means we don't support the operation at all. Invalid operation means we support in general but the the operation is invalid given the current arguments and/or state of the object.

Reply:
sgtm. Done